pose_lookup/plugin.cpp

```diff
@@ -5,138 +5,185 @@
+        init_pos_offset = _m_sensor_data.cbegin()->second.position;
+
+        auto newoffset = correct_pose(_m_sensor_data.begin()->second).orientation;
+        set_offset(newoffset);
     }
 
     virtual fast_pose_type get_fast_pose() const override {
-		const time_type* estimated_vsync = _m_vsync_estimate->get_latest_ro();
-		if(estimated_vsync == nullptr) {
-			std::cerr << "Vsync estimation not valid yet, returning fast_pose for now()" << std::endl;
-			return get_fast_pose(std::chrono::system_clock::now());
-		} else {
-			return get_fast_pose(*estimated_vsync);
-		}
+        const time_type* estimated_vsync = _m_vsync_estimate->get_latest_ro();
+        if(estimated_vsync == nullptr) {
+            std::cerr << "Vsync estimation not valid yet, returning fast_pose for now()" << std::endl;
+            return get_fast_pose(std::chrono::system_clock::now());
+        } else {
+            return get_fast_pose(*estimated_vsync);
+        }
     }
 
     virtual pose_type get_true_pose() const override {
-		throw std::logic_error{"Not Implemented"};
+        throw std::logic_error{"Not Implemented"};
     }
 
-
     virtual bool fast_pose_reliable() const override {
-		return true;
+        return true;
     }
 
     virtual bool true_pose_reliable() const override {
-		return false;
+        return false;
     }
 
-	virtual Eigen::Quaternionf get_offset() override {
+    virtual Eigen::Quaternionf get_offset() override {
         return offset;
     }
 
-	virtual pose_type correct_pose(const pose_type pose) const override {
-		pose_type swapped_pose;
-
-		// This uses the OpenVINS standard output coordinate system.
-		// This is a mapping between the OV coordinate system and the OpenGL system.
-		swapped_pose.position.x() = -pose.position.y();
-		swapped_pose.position.y() = pose.position.z();
-		swapped_pose.position.z() = -pose.position.x();
-
-		// There is a slight issue with the orientations: basically,
-		// the output orientation acts as though the "top of the head" is the
-		// forward direction, and the "eye direction" is the up direction.
-		Eigen::Quaternionf raw_o (pose.orientation.w(), -pose.orientation.y(), pose.orientation.z(), -pose.orientation.x());
-
-		swapped_pose.orientation = apply_offset(raw_o);
-
-		return swapped_pose;
-	}
+    virtual pose_type correct_pose(const pose_type pose) const override {
+        pose_type swapped_pose;
+        pose_type input_pose;
+
+        // Step 1: Compensate starting point to (0, 0, 0), pos only
+        input_pose.position(0) = pose.position(0) - init_pos_offset(0);
+        input_pose.position(1) = pose.position(1) - init_pos_offset(1);
+        input_pose.position(2) = pose.position(2) - init_pos_offset(2);
+        input_pose.orientation.x() = pose.orientation.x();
+        input_pose.orientation.y() = pose.orientation.y();
+        input_pose.orientation.z() = pose.orientation.z();
+        input_pose.orientation.w() = pose.orientation.w();
+
```

Comment:
I'm not bothered by the extra copying here as `input_pose` is a temporary local. (@charmoniumQ @mhuzai) Thoughts?

Reply:
Use the functional-style transformation for this part now.